In [1]:
import pandas as pd

# The DataFrame Data Structure

A **DataFrame** is the primary object that we'll be using for cleaning and analyzing structured data sets. We can think of a DataFrame as essentially a 2-dimensional Series, or as a table with labeled rows and columns.

There are many ways to create a DataFrame. For example, we can use a group of series, where each Series represents a row of data. We could also use a group of dictionaries, where each dictionary represents a row of data. For now, let's use the first method.

In [2]:
# Three Series with the same indexes
purchase_1 = pd.Series({'Name': 'Mr. Chung',
                        'Item': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Dr. Gaines',
                        'Item': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Mr. Lucey',
                        'Item': 'Bird Seed',
                        'Cost': 5.00})

# Combine Series into a DataFrame, setting index values indicating where each purchase occurred
df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])
df

Name          Item  Cost
Store 1   Mr. Chung      Dog Food  22.5
Store 1  Dr. Gaines  Kitty Litter   2.5
Store 2   Mr. Lucey     Bird Seed   5.0

## Extracting data
We can extract data from the DataFrame using the `loc` and `iloc` attributes, just like with a Series. When we use these attributes on a DataFrame, we query for a specific *row* as given by the index label.

In [4]:
# Rows: Store 2
# Cols: (all)
df.loc['Store 2']

Name    Mr. Lucey
Item    Bird Seed
Cost            5
Name: Store 2, dtype: object

Note that the object returned is a Series, since there was only one row corresponding to the index `Store 2`. We can verify this using the Python `type()` function.

In [5]:
type(df.loc['Store 2'])

pandas.core.series.Series

If there are multiple matching rows, then the result is returned as another DataFrame.

In [6]:
# Rows: Store 1
# Cols: (all)
df.loc['Store 1']

Name          Item  Cost
Store 1   Mr. Chung      Dog Food  22.5
Store 1  Dr. Gaines  Kitty Litter   2.5

The DataFrame object allows us to easily select data based on multiple axes. For example, if we want to list all of the costs for `Store 1`, all we have to do is supply two parameters to the `loc` attribute: the row name and the column name.

In [7]:
# Rows: Store 1
# Cols: Cost
df.loc['Store 1', 'Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

What if we want to select an entire column, such as all of the costs? This may seem tricky, because every `loc` or `iloc` query must start with a row index. However, Pandas provides some ways around this.

The first way takes advantage of the **transpose** of the DataFrame, which is what we call the same data with the rows and columns flipped.

In [8]:
df.T

Store 1       Store 1    Store 2
Name  Mr. Chung    Dr. Gaines  Mr. Lucey
Item   Dog Food  Kitty Litter  Bird Seed
Cost       22.5           2.5          5

By taking the transpose of `df`, we make it so that `Cost` is its own row; then, we can simply query using the `loc` attribute as usual.

In [9]:
# Rows: Cost
# Cols: (all)
df.T.loc['Cost']

Store 1    22.5
Store 1     2.5
Store 2       5
Name: Cost, dtype: object

However, this is a little bit ugly, so Pandas provides an even easier way! Since `loc` and `iloc` query by row, the operation of indexing directly on a DataFrame is reserved for querying by column. Columns always have names, so this will never cause a problem like it might with a Series.

In [10]:
# Rows: (all)
# Cols: Cost
df['Cost']

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

Since querying a DataFrame returns either a Series or another DataFrame, we can **chain** operations together. Here's another way we could have queried for all `Store 1` costs, using operation chaining.

In [11]:
# Rows: Store 1
# Cols: Cost
df.loc['Store 1']['Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

**Important note:** chaining is great, but be careful about when you use it! Chaining causes Pandas to return a *copy* of the DataFrame instead of a *view* of the original, and is thus slower and more memory-intensive than necessary.

For the purposes of this course, the larger issue will be the fact that chaining returns a copy. In the future, when we start changing data values, it's important that we know the difference between editing a DataFrame and editing a copy.

## Selecting and Slicing
As we saw before, `loc` can take two parameters: a row index and a column name. We can also provide a list of column names to select. For example, let's select the `Name` and `Cost` for every purchase at `Store 1`.

In [12]:
# First, let's report df again to remind ourselves of what it looks like
df

Name          Item  Cost
Store 1   Mr. Chung      Dog Food  22.5
Store 1  Dr. Gaines  Kitty Litter   2.5
Store 2   Mr. Lucey     Bird Seed   5.0

In [16]:
# Rows: Store 1
# Cols: Name, Cost
df.loc['Store 1', ['Name', 'Cost']]

Name  Cost
Store 1   Mr. Chung  22.5
Store 1  Dr. Gaines   2.5

The `loc` operator also supports slicing, for both rows and columns. For example, if we wanted to select the `Name` and `Cost` for every row, we can do the following:

In [19]:
# Rows: (all)
# Cols: Name, Cost
df.loc[:, ['Name', 'Cost']]

Name  Cost
Store 1   Mr. Chung  22.5
Store 1  Dr. Gaines   2.5
Store 2   Mr. Lucey   5.0

As one more example of slicing, let's select all rows, but only the columns starting from `Item` and moving to the right (a.k.a. `Item` and `Cost`).

In [22]:
# Rows: (all)
# Cols: Item, Cost
df.loc[:, 'Item':]

Item  Cost
Store 1      Dog Food  22.5
Store 1  Kitty Litter   2.5
Store 2     Bird Seed   5.0

## Dropping and adding data
To delete data from a DataFrame (or from a Series), we use the `drop()` function. But before we do that, let's remind ourselves of what our original DataFrame looks like, so that we can see the difference after dropping data.

In [23]:
df

Name          Item  Cost
Store 1   Mr. Chung      Dog Food  22.5
Store 1  Dr. Gaines  Kitty Litter   2.5
Store 2   Mr. Lucey     Bird Seed   5.0

To remove a row, we simply call `drop()` and pass in the label of the row to remove.

In [24]:
# Drop the row specified by index label
df.drop('Store 1')

Name       Item  Cost
Store 2  Mr. Lucey  Bird Seed   5.0

Note that the `drop()` method does not change the original DataFrame, but rather returns a copy in which the specified rows have been removed. We can verify this by taking another look at our original DataFrame – nothing has changed.

In [25]:
df

Name          Item  Cost
Store 1   Mr. Chung      Dog Food  22.5
Store 1  Dr. Gaines  Kitty Litter   2.5
Store 2   Mr. Lucey     Bird Seed   5.0

To store the copy with removed rows, we must store the result in a new variable.

In [36]:
# Store reduced view in a new variable
df_view = df.drop('Store 1')
df_view

Name       Item  Cost
Store 2  Mr. Lucey  Bird Seed   5.0

We can remove a specified column by passing in an optional parameter `axis=1`, which tells Pandas that the label should apply to columns instead of rows.

In [3]:
# Drop a column by specifying axis of operation
df_view2 = df.drop('Item', axis=1)
df_view2

Name  Cost
Store 1   Mr. Chung  22.5
Store 1  Dr. Gaines   2.5
Store 2   Mr. Lucey   5.0

Another way to remove a column is to index the DataFrame directly (which, remember, returns a column) and combine it with the `del` keyword. This will change the DataFrame directly without creating a copy, but it will not return a view of the updated data.

In [37]:
# Delete a column directly using del
del df_view['Name']
df_view

Item  Cost
Store 2  Bird Seed   5.0

Finally, we can easily add a new column to a DataFrame. For example, if we want to add a new column `Location` with default values of `None`, we can do it as follows. Note that the value `None` gets broadcasted across all rows.

In [4]:
# Add a new column with default values of None
df['Location'] = None
df

Name          Item  Cost Location
Store 1   Mr. Chung      Dog Food  22.5     None
Store 1  Dr. Gaines  Kitty Litter   2.5     None
Store 2   Mr. Lucey     Bird Seed   5.0     None

If we wish to specify a separate value for each row, we can do that using a list of the values to assign.

In [5]:
# Add a new column with specific values
df['Location'] = ['Brooklyn', 'Brooklyn', 'Manhattan']
df

Name          Item  Cost   Location
Store 1   Mr. Chung      Dog Food  22.5   Brooklyn
Store 1  Dr. Gaines  Kitty Litter   2.5   Brooklyn
Store 2   Mr. Lucey     Bird Seed   5.0  Manhattan

# DataFrame Indexing and Loading

A common workflow is to read your data into a DataFrame, and then reduce that DataFrame into the specific columns or rows that you're interested in working with.

## DataFrame views vs. copies
When used properly, DataFrame queries give you a *view* of the data being accessed. This is much faster and more memory efficient than copying data. But this also means that you will have to be careful to keep track of which types of operations change a DataFrame in place, so that you always know the state of the DataFrame at any given point in time.

Here's an example of an operation that changes the underlying DataFrame. Using the same purchasing data as before, we create a Series out of the `Cost` column.

In [6]:
costs = df['Cost']
costs

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

Now, we use broadcasting to increment every value in the Series by 2.

In [7]:
costs += 2
costs

Store 1    24.5
Store 1     4.5
Store 2     7.0
Name: Cost, dtype: float64

What you might not expect is that this actually changes the values in the original DataFrame as well. This is because querying the `Cost` column out of the DataFrame returned a *view* (a.k.a. not a copy) of the original, so any changes applied to that view are also applied to the underlying DataFrame.

We can verify that the change has happened to the original data by taking another look at it.

In [8]:
df

Name          Item  Cost   Location
Store 1   Mr. Chung      Dog Food  24.5   Brooklyn
Store 1  Dr. Gaines  Kitty Litter   4.5   Brooklyn
Store 2   Mr. Lucey     Bird Seed   7.0  Manhattan

In [9]:
store1_costs = df.loc['Store 1', 'Cost']
store1_costs += 100
df

Name          Item  Cost   Location
Store 1   Mr. Chung      Dog Food  24.5   Brooklyn
Store 1  Dr. Gaines  Kitty Litter   4.5   Brooklyn
Store 2   Mr. Lucey     Bird Seed   7.0  Manhattan

## Reading and indexing CSV files

We'll be working a lot with moderately-sized data sets, which are often stored externally in what we call "Comma-Separated Values" (CSV) files. In order to work with such a data set, we must first read the file into a DataFrame.

Here's an example of [a CSV file](https://raw.githubusercontent.com/michaelschung/bc-data-processing/master/datasets/zoo.csv), which contains data summarizing the amount of water needed by each animal in a zoo.

Let's read this CSV file into a DataFrame.

In [12]:
zoo_url = 'zoo.csv'

zoo = pd.read_csv(zoo_url, index_col=0)
zoo

uniq_id  water_need
animal                       
elephant     1001         500
elephant     1002         600
elephant     1003         550
tiger        1004         300
tiger        1005         320
tiger        1006         330
tiger        1007         290
tiger        1008         310
zebra        1009         200
zebra        1010         220
zebra        1011         240
zebra        1012         230
zebra        1013         220
zebra        1014         100
zebra        1015          80
lion         1016         420
lion         1017         600
lion         1018         500
lion         1019         390
kangaroo     1020         410
kangaroo     1021         430
kangaroo     1022         410

If we want to get an idea of what our DataFrame looks like but we don't want to report the entire thing, we can use `zoo.head()` to display the first five rows.

In [68]:
# Use zoo.head() to display first five rows
zoo.head()

uniq_id  water_need
animal                       
elephant     1001         500
elephant     1002         600
elephant     1003         550
tiger        1004         300
tiger        1005         320

We can also pass a number into `zoo.head()` to specify a specific number of rows to report.

In [71]:
# Specify a specific number of rows to display
zoo.head(10)

uniq_id  water_need
animal                       
elephant     1001         500
elephant     1002         600
elephant     1003         550
tiger        1004         300
tiger        1005         320
tiger        1006         330
tiger        1007         290
tiger        1008         310
zebra        1009         200
zebra        1010         220

Note that currently, the index of our DataFrame is given by default integer values. However, in this case this is unnecessary, because every animal already has a unique ID number, specified in the `uniq_id` column.

Instead, it would be better for the `animal` column to be the index, so that we can query for all elephants, all tigers, etc. We accomplish this using the `set_index()` method, which takes a column name as a specifier makes it the new index.

In [ ]:
# Set index column


_Note that `zoo.set_index('animal')` does not change the `zoo` DataFrame on its own. So to avoid making unnecessary copies, we are overwriting the previous `zoo` with its new copy with an updated index._

An alternative way to set the index is to do so during the `read_csv` process. To accomplish this, we use an extra parameter `index_col` to tell Pandas ahead of time which column to use as the index.

In [ ]:
# Use the 0th column as the index while reading in the CSV


# Querying a DataFrame

Let's say we want to find out which animals need more than 300 gallons of water. The column name for amount of water needed is `water_need`, so we can do the following:

In [72]:
# Ask for animals that need more than 300 gallons of water
zoo['water_need'] > 300

animal
elephant     True
elephant     True
elephant     True
tiger       False
tiger        True
tiger        True
tiger       False
tiger        True
zebra       False
zebra       False
zebra       False
zebra       False
zebra       False
zebra       False
zebra       False
lion         True
lion         True
lion         True
lion         True
kangaroo     True
kangaroo     True
kangaroo     True
Name: water_need, dtype: bool

Note the format of the output here: we are given a Series consisting of a bunch of Booleans. We call this a **Boolean mask**, because we can use it to "mask" the results of the DataFrame.

For example, perhaps we want to see _all_ of the information for each animal that requires more than 300 gallons of water. We can use our Boolean mask to query the DataFrame using the `where()` method.

In [74]:
# Use Boolean mask to query the DataFrame for all information on guzzlers
guzzlers = zoo.where(zoo['water_need'] > 300)
guzzlers

uniq_id  water_need
animal                       
elephant   1001.0       500.0
elephant   1002.0       600.0
elephant   1003.0       550.0
tiger         NaN         NaN
tiger      1005.0       320.0
tiger      1006.0       330.0
tiger         NaN         NaN
tiger      1008.0       310.0
zebra         NaN         NaN
zebra         NaN         NaN
zebra         NaN         NaN
zebra         NaN         NaN
zebra         NaN         NaN
zebra         NaN         NaN
zebra         NaN         NaN
lion       1016.0       420.0
lion       1017.0       600.0
lion       1018.0       500.0
lion       1019.0       390.0
kangaroo   1020.0       410.0
kangaroo   1021.0       430.0
kangaroo   1022.0       410.0

You can think of that command as basically saying, "Give me all of the rows in the DataFrame where the `water_need` column has a value greater than 300."

You may notice that for every animal with a lower water  need, its entire row was filled with `NaN` (Python's abbreviation for "not a number"). That's okay! This is Pandas' way of telling you that those rows are not applicable given the query you just performed.

These `NaN`s may look annoying and problematic, but Pandas knows to ignore them when it performs any further operations. We can show this by comparing the number of rows in the `water_need` column of `guzzlers` to the analogous number in our original table `zoo`.

In [80]:
print(zoo.count())
print(guzzlers.count())

uniq_id       22
water_need    22
dtype: int64
uniq_id       13
water_need    13
dtype: int64


Of course, it'll usually be nicer to drop the `NaN` rows anyway, and we can do that using the `dropna()` method. Recall that the `drop()` method (discussed earlier) returns a copy of the original DataFrame. Similarly, `dropna()` by default also returns a copy, so we must include the parameter `inplace=True` to force `guzzlers` to apply the change to itself.

_Alternatively, we could omit the extra parameter and just update `guzzlers` to the copy of itself - similar to what we did when we changed the index._

In [81]:
# Drop NaN rows
guzzlers.dropna(inplace=True)
guzzlers

uniq_id  water_need
animal                       
elephant   1001.0       500.0
elephant   1002.0       600.0
elephant   1003.0       550.0
tiger      1005.0       320.0
tiger      1006.0       330.0
tiger      1008.0       310.0
lion       1016.0       420.0
lion       1017.0       600.0
lion       1018.0       500.0
lion       1019.0       390.0
kangaroo   1020.0       410.0
kangaroo   1021.0       430.0
kangaroo   1022.0       410.0

So let's summarize. To query specific rows from a DataFrame, we need to:

1. Create a Boolean mask based on column values in question
2. Pass this mask into the `where()` function of the DataFrame
3. Remove `NaN` rows using `dropna()`

Let's try one more example together. Write a query for all animals that have ID numbers greater than or equal to 1010.

In [85]:
mask = zoo['uniq_id'] >= 1010
newbies = zoo.where(mask)
newbies.dropna(inplace=True)
newbies

uniq_id  water_need
animal                       
zebra      1010.0       220.0
zebra      1011.0       240.0
zebra      1012.0       230.0
zebra      1013.0       220.0
zebra      1014.0       100.0
zebra      1015.0        80.0
lion       1016.0       420.0
lion       1017.0       600.0
lion       1018.0       500.0
lion       1019.0       390.0
kangaroo   1020.0       410.0
kangaroo   1021.0       430.0
kangaroo   1022.0       410.0

# Advanced Querying

## Shorthand queries

Now that you've mastered the pattern of querying, it's time you know that there's actually a faster way. Querying for a specific subset and dropping the `NaN` rows is such a common set of actions that Pandas provies a shorthand way to do it. All we need to do is to query the DataFrame directly, using a Boolean mask instead of an index label.

In [86]:
# Shortcut way to query!
zoo[zoo['water_need'] > 300]

uniq_id  water_need
animal                       
elephant     1001         500
elephant     1002         600
elephant     1003         550
tiger        1005         320
tiger        1006         330
tiger        1008         310
lion         1016         420
lion         1017         600
lion         1018         500
lion         1019         390
kangaroo     1020         410
kangaroo     1021         430
kangaroo     1022         410

If that looks confusing, think of it this way:

In [87]:
# Create a mask
mask = zoo['water_need'] > 300
# Use the mask to query the DataFrame directly
zoo[mask]

uniq_id  water_need
animal                       
elephant     1001         500
elephant     1002         600
elephant     1003         550
tiger        1005         320
tiger        1006         330
tiger        1008         310
lion         1016         420
lion         1017         600
lion         1018         500
lion         1019         390
kangaroo     1020         410
kangaroo     1021         430
kangaroo     1022         410

## Combining Boolean masks

We can combine multiple Boolean masks to accomplish more complex queries.

Previously, we queried to see which animals require more than 300 gallons of water, as well as which have ID numbers greater than or equal to 1010. Now, let's query for which animals fulfill both of those at the same time. We can do this by combining the two masks using the `&` operator.

In [13]:
guzzler_mask = zoo['water_need'] > 300
geq_1010 = zoo['uniq_id'] >= 1010
zoo[guzzler_mask & geq_1010]

uniq_id  water_need
animal                       
lion         1016         420
lion         1017         600
lion         1018         500
lion         1019         390
kangaroo     1020         410
kangaroo     1021         430
kangaroo     1022         410

Of course, all of that code can be condensed into one single line.

**Note: when Boolean masks are combined inline like this, each mask _must_ be encased in parentheses.**

In [16]:
# Parentheses around each Boolean mask!
zoo[(zoo['water_need'] > 300) & (zoo['uniq_id'] >= 1010)]

uniq_id  water_need
animal                       
lion         1016         420
lion         1017         600
lion         1018         500
lion         1019         390
kangaroo     1020         410
kangaroo     1021         430
kangaroo     1022         410

Now, write a query that finds all animals with ID numbers less than 1014, which also require less than 250 gallons of water.

As a final exercise, write a query that finds all animals with ID numbers from 1010-1019, and also require between 200-399 gallons of water.

In [19]:
id_range = (zoo['uniq_id'] >= 1010) & (zoo['uniq_id'] < 1020)
water_range = (zoo['water_need'] >= 200) & (zoo['water_need'] < 400)

zoo[id_range & water_range]

uniq_id  water_need
animal                     
zebra      1010         220
zebra      1011         240
zebra      1012         230
zebra      1013         220
lion       1019         390